tmpl/wiki/tutorial-lwt.md

```diff
@@ -74,508 +111,327 @@ Assuming you are building for POSIX using native kernel sockets, compile the app
+##Challenge 2: Echo server
 
-This is a good place to introduce some of these extensions. When opening the Lwt module, the infix operator `>>=` is made available. This operator is an alternative to the `bind` function and often makes the code more readable. E.g. consider `bind (bind (bind t f) g) h` and the operator based equivalent expression `t >>= f >>= g >>= h`. We can now rewrite the previous solution more simply:
+Write an echo server that reads from a dummy input generator and, for each line it reads, writes it to the console. The server should stop after reading 10 lines. Here is a basic dummy input generator:
 
 ```
-  open V1_LWT (* provides the CONSOLE signature *)
-  open Lwt    (* provides >>= and join *)
-  open OS     (* provides Env, Main, and Time *)
-
-  module Main (C : CONSOLE) = struct
-    let start c =
-      join [
-        (Time.sleep 1.0 >>= fun () -> (C.log c "Heads"; return ()));
-        (Time.sleep 2.0 >>= fun () -> (C.log c "Tails"; return ()));
-       ] >>= fun () ->
-         C.log c "Finished";
-         return ()
-  end
+  let read_line () =
+    OS.Time.sleep (Random.float 2.5) >|= fun () ->
+    String.make (Random.int 20) 'a'
 ```
 
-This is in `lwt/src/unikernels.ml` in the repository. The target is `heads2`.
-
-
-###Anonymous Bind
-
-If you are chaining sequences of blocking I/O, a common pattern is to write:
+###Solution
 
 ```
-  write stdio "Foo" >>= fun () ->
-  write stdio "Bar" >>= fun () ->
-  write stdio "Done"
+  let start c =
+    let rec echo_server = function
+      | 0 -> Lwt.return ()
+      | n ->
+        read_line () >>= fun s ->
+        C.log_s c s >>= fun () ->
+        echo_server (n - 1)
+    in
+    echo_server 10
 ```
 
-You can replace these anonymous binds with the `>>` operator instead:
+This is in `lwt/src/unikernels.ml` in the repository. Build with the target `echo_server1`.
 
-```
-  write stdio "Foo" >>
-  write stdio "Bar" >>
-  write stdio "Done"
-```
 
-###Lwt Bindings
+##The main event loop
 
-The binding operation reverses the normal `let` binding by specifying the name of the bound variable in the second argument. Consider the thread:
+Understanding the basic principles behind Lwt can be helpful.
 
-```
-  e1 >>= fun x -> e2
-```
+The core of Lwt is based on an event loop. In "standard" (non-MirageOS) settings,
+this loop is started using the `Lwt_main.run` function. However, when using
+MirageOS, the loop is automatically started by the `main.ml` file autogenerated
+by the `mirage` command-line tool.
 
-Here, we wait for the result of `e1`, bind the result to `x` and continue into `e2`. You can replace this with the more natural `lwt` syntax to act as a "blocking let":
+Because it's based on an event loop, threads are actually very cheap in Lwt
+when compared to preemptive system threads. Sleeping actually registers an
+event that will wake up the associated thread when possible. Depending on the
+backend, the event registering slightly differs.
```

Comment:
The last sentence worries me – do I have to worry about this difference as a user? I think it's better just to drop it.